gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientReadChannel.java

```diff
@@ -0,0 +1,505 @@
+
+  private boolean shouldDetectRandomAccess() {
+    return !randomAccess && readOptions.getFadvise() == Fadvise.AUTO;
+  }
+
+  private boolean isRandomAccessPattern(long oldPosition) {
+    if (!shouldDetectRandomAccess()) {
+      return false;
+    }
+    // backwardSeek
+    if (currentPosition < oldPosition) {
+      logger.atFine().log(
+          "Detected backward read from %s to %s position, switching to random IO for '%s'",
+          oldPosition, currentPosition, resourceId);
+      return true;
+    }
+    if (oldPosition >= 0 && oldPosition + readOptions.getInplaceSeekLimit() < currentPosition) {
+      logger.atFine().log(
+          "Detected forward read from %s to %s position over %s threshold,"
+              + " switching to random IO for '%s'",
+          oldPosition, currentPosition, readOptions.getInplaceSeekLimit(), resourceId);
+      return true;
+    }
+    return false;
+  }
+
+  private void skipInPlace(long seekDistance) {
+    if (skipBuffer == null) {
+      skipBuffer = new byte[SKIP_BUFFER_SIZE];
+    }
+    while (seekDistance > 0 && contentReadChannel != null) {
+      try {
+        int bufferSize = toIntExact(min(skipBuffer.length, seekDistance));
+        int bytesRead = contentReadChannel.read(ByteBuffer.wrap(skipBuffer, 0, bufferSize));
+        if (bytesRead < 0) {
+          logger.atInfo().log(
+              "Somehow read %d bytes trying to skip %d bytes to seek to position %d, size: %d",
+              bytesRead, seekDistance, currentPosition, objectSize);
+          closeContentChannel();
+        } else {
+          seekDistance -= bytesRead;
+          contentChannelPosition += bytesRead;
+        }
+      } catch (IOException e) {
+        logger.atInfo().withCause(e).log(
+            "Got an IO exception on contentChannel.read(), a lazy-seek will be pending for '%s'",
+            resourceId);
+        closeContentChannel();
+      }
+    }
+    checkState(
+        contentReadChannel == null || contentChannelPosition == currentPosition,
+        "contentChannelPosition (%s) should be equal to currentPosition (%s)"
+            + " after successful in-place skip",
+        contentChannelPosition,
+        currentPosition);
+  }
+
+  private void resetContentChannel() {
+    checkState(contentReadChannel == null, "contentChannel should be null for '%s'", resourceId);
+    contentChannelPosition = -1;
+    contentChannelEnd = -1;
+  }
+
+  protected void closeContentChannel() {
+    if (contentReadChannel != null) {
+      logger.atFiner().log("Closing internal contentChannel for '%s'", resourceId);
+      try {
+        contentReadChannel.close();
+      } catch (Exception e) {
+        logger.atFine().withCause(e).log(
+            "Got an exception on contentChannel.close() for '%s'; ignoring it.", resourceId);
+      } finally {
+        contentReadChannel = null;
+        resetContentChannel();
+      }
+    }
+  }
+
+  private void performPendingSeeks(long bytesToRead) throws IOException {
```

Comment:
There is a lot of code duplication with GoogleCloudStorageReadChannel. Please refactor and reuse.

Reply:
I do understand that their is a lot of code duplication but I don't see a better approach as a lot of channel pointers i.e. contentChannelPosition, currentPosition etc. are getting modified at multiple places. 

I will try my best to refactor and put some validators in Helper class but we will still have to live with some code duplication. Will discuss more offline.

Reply:
We are having logic similar to the existing code. Can we try to see if there is any opportunity to make the code simpler and make full use of the new library.